pdr_backend/trueval/trueval_agent.py

```diff
@@ -0,0 +1,148 @@
+            print(f"No pending slots, sleeping for {self.config.sleep_time} seconds...")
+            time.sleep(self.config.sleep_time)
+            return
+
+        for slot in pending_slots:
+            print("-" * 30)
+            print(f"Processing slot {slot.slot} for contract {slot.contract.address}")
+            try:
+                self.process_slot(slot)
+            except Exception as e:
+                print("An error occured", e)
+        print(f"Done processing, sleeping for {self.config.sleep_time} seconds...")
+        time.sleep(self.config.sleep_time)
+
+    @enforce_types
+    def process_slot(self, slot: Slot) -> dict:
+        predictoor_contract, seconds_per_epoch = self.get_contract_info(
+            slot.contract.address
+        )
+        return self.get_and_submit_trueval(slot, predictoor_contract, seconds_per_epoch)
+
+    @enforce_types
+    def get_contract_info(
+        self, contract_address: str
+    ) -> Tuple[PredictoorContract, int]:
+        if contract_address in self.contract_cache:
+            predictoor_contract, seconds_per_epoch = self.contract_cache[
+                contract_address
+            ]
+        else:
+            predictoor_contract = PredictoorContract(
+                self.config.web3_config, contract_address
+            )
+            seconds_per_epoch = predictoor_contract.get_secondsPerEpoch()
+            self.contract_cache[contract_address] = (
+                predictoor_contract,
+                seconds_per_epoch,
+            )
+        return (predictoor_contract, seconds_per_epoch)
+
+    def get_and_submit_trueval(
+        self,
+        slot: Slot,
+        predictoor_contract: PredictoorContract,
+        seconds_per_epoch: int,
+    ) -> dict:
+        slot.slot = int(slot.slot)
+        initial_ts = slot.slot - seconds_per_epoch
+        end_ts = slot.slot
+
+        (true_val, error) = self.get_true_val(slot.contract, initial_ts, end_ts)
+        if error:
+            raise Exception(
+                f"Error getting trueval for {slot.contract.pair} and slot {slot.slot}"
+            )
+
+        # pylint: disable=line-too-long
+        print(
+            f"Contract:{predictoor_contract.contract_address} - Submitting true_val {true_val} and slot:{slot.slot}"
+        )
+
+        tx = predictoor_contract.submit_trueval(true_val, slot.slot, False, True)
+
+        return tx
+
+
+def get_true_val(
+    topic: ContractData, initial_timestamp: int, end_timestamp: int
+) -> Tuple[bool, bool]:
+    """Given a topic, Returns the true val between end_timestamp and initial_timestamp
+    Topic object looks like:
+
+    {
+        "name":"ETH-USDT",
+        "address":"0x54b5ebeed85f4178c6cb98dd185067991d058d55",
+        "symbol":"ETH-USDT",
+        "blocks_per_epoch":"60",
+        "blocks_per_subscription":"86400",
+        "pair":"eth-usdt",
+        "base":"eth",
+        "quote":"usdt",
+        "source":"kraken",
+        "timeframe":"5m"
+    }
+
```

Comment:
You don't need to specify how the "topic" (feed) data is shaped, since that info is in ContractData